### DISCUSSION TITLE
Legal numbering

### DISCUSSION THREAD

**Discussion by maradotwebp** [typst/typst#2506] [Show and tell]
I am using Typst to write a legal document, and had the need to number paragraphs so that they can be referenced more easily.

I thought I'd share my solution here. It changes all occurences of numbered lists to be prefixed with the numbering of the last header, although I'm sure you could adapt the code if you want a custom function instead of `enum`.

```
// Legal formatting of enumeration
show enum: it => locate(loc => {
  // Retrieve the last heading so we know what level to step at
  let headings = query(selector(heading).before(loc), loc)
  let last = headings.at(-1)
  
  // Combine the output items
  let output = ()
  for item in it.children {
    output.push([
      #counter(heading).step(level: last.level + 1)
      #counter(heading).display()
    ])
    output.push([
      #text(item.body)
      #parbreak()
    ])
  }
  
  // Display in a grid
  grid(
    columns: (auto, 1fr),
    column-gutter: 1em,
    row-gutter: 1em,
    ..output
  )
})
```

Altogether, it makes this `.typ` code:

```
= Some heading
+ Paragraphs here are preceded with a number so they can be referenced directly.
+ _#lorem(100)_
+ _#lorem(100)_

== A subheading
+ Paragraphs are also numbered correctly in subheadings.
+ _#lorem(50)_
+ _#lorem(50)_
```

look like this:

![Output](https://i.imgur.com/XF5is6j.png)

**Comment by ckunte** [comment]
Hi, thanks for this script. However, it can't seem to get it to work in Typst 0.11.1?

**Reply by rriemann** [reply]
The example has been cross-posted to https://sitandr.github.io/typst-examples-book/book/snippets/numbering.html?highlight=para#numbering-each-paragraph

I copied it from there it works for me with version 0.11.1

**Reply by ckunte** [reply]
Thank you @rriemann; much appreciate the share. Works for me.

**Comment by kenjknox** [comment]
There's an alternative approach below:

```typst

#show heading: element => {
  place(
    dx:-1cm,
    context {counter(heading).display()}
  )
  element.body
}

#show regex("^\| "): it => context {
  let headings = query(selector(heading).before(here()))
  let last = headings.at(-1)
  let output = {
    context {counter(heading).step(level: last.level + 1)}
    context {counter(heading).display()}
  }
  place(
    dx: -1cm,
    output
  )
}

```

![image](https://github.com/user-attachments/assets/82dbd6af-98bd-4ad7-adc4-f520ce163f9d)

Very much the same idea, but have moved the numbering into the margin and assigned the | character to mark the start line of each clause. Idea is that doing it this way means you're not overriding the enum function and not having to use a grid to format. Does mean you have a fixed offset, so need to be careful you leave enough room.

**Comment by DaHaiz** [comment]
I've run into problems with the solution of @kenjknox when a paragraph would be pushed to the next page but the numbering would stuck on the previous one. I ended up using a combined solution of @kenjknox (not using enums) and @maradotwebp (grids, but grid per heading and per par).

This has also the advantage that the numbering is left aligned but still in the flow. No margin usage therefore and no need to take care about "absolute" placements.

```
#show heading: element => {
  grid(
      columns: (1cm, 1fr),
      column-gutter: 1em,
      row-gutter: 1em,
      context {counter(heading).display()},
      block(element.body)
    )
}

#show regex("^\| .*"): it => context {
  let para = it.text.trim("| ")
  let headings = query(selector(heading).before(here()))
  let last = headings.at(-1)
  let output = {
    context {counter(heading).step(level: last.level + 1)}
    context {counter(heading).display()}
  }

  grid(
      columns: (1cm, 1fr),
      column-gutter: 1em,
      row-gutter: 1em,
      output,
      para
    )
    
}
```

hope this helps someone :)

**Reply by kenjknox** [reply]
I eventually ran into this problem myself, so would definitely recommend your solution now. Think you're right about it being better keeping the numbering in flow. Will be treating offsets with more care in future.

**Comment by cschreib-ibex** [comment]
I think there's an easier way in 0.12.0:
```typst
#set enum(
    number-align: alignment.left,
    numbering: n => context {
    let headings = query(selector(heading).before(here()))
    let last = headings.at(-1)
    counter(heading).step(level: last.level + 1)
    context {counter(heading).display()}
})

#set heading(numbering: "1.")
```

![image](https://github.com/user-attachments/assets/c51d9991-9b87-438a-a533-f53431b42cc5)


The other solutions with the `#show regex` start failing when displaying inline raw text, e.g.:
```typst
= Some heading

| Paragraphs can include `raw code` and continue seamlessly
```
![image](https://github.com/user-attachments/assets/4e8ef32a-d2a0-4fb2-8a1d-808a0e791cfb)

**Reply by madchap** [reply]
That one worked straight OOTB for me, thanks!

**Reply by madchap** [reply]
I do have an issue however if try to sub-bullet, e.g.:

```
= Services and payments

+ Services. The Consultant shall provide the Services to the Company as specified in one or more Statements of Work. Each Statement of Work shall include, at a minimum, a description of the Services to be performed, the Deliverables to be provided, the timeline for completion, and the fees and payment schedule.
+ Performance Standard. The Consultant shall perform the Services in a professional and workmanlike manner, in accordance with industry standards and practices, and shall comply with all applicable laws, regulations, and professional standards.
+ Payment. As full consideration for the performance of the Services, the Company shall pay the Consultant the fees specified in the applicable Statement of Work. Unless otherwise specified in a Statement of Work:
  + The Consultant shall invoice the Company monthly for Services performed;
```

That last bullet point is causing an error

```
error: unexpected argument
   ┌─ consulting-agreements/main.typ:19:15
   │
19 │     numbering: n => context {
   │                ^
```

**Reply by maradotwebp** [reply]
@madchap Iterating on the above solution, this is the code I came up with.
It includes the level of the previous enum item for sub lists.
**`full: true` must be set for this to work.**

```typst
#set heading(numbering: "1.")

#set enum(
  full: true,
  numbering: (..n) => context {
    let headings = query(selector(heading).before(here()))
    let last = headings.at(-1)
    counter(heading).step(level: last.level + n.pos().len())
    context {counter(heading).display()}
  }
)
```

![legal_lining](https://github.com/user-attachments/assets/212418a9-fcb2-44cc-897d-30198e3f8561)

with the following code:

```typst
= Heading \#1

+ #lorem(50)
+ #lorem(50)
+ #lorem(50)

== Subheadings

+ #lorem(50)
  + #lorem(50)
    + #lorem(50)
    + #lorem(50)
  + #lorem(50)
+ #lorem(50)

=== Heading \#2

+ #lorem(20)
```

works in Typst `0.11.0` and above.

**Reply by madchap** [reply]
Very neat, you're a rockstar :-D thanks a lot!
